J2735-2024 Part 7 of 9: RUCRM, TUM, TUMack, TLSM, Test & future use messages

iyourshaw opened the pull request:
# Part 7

## New Message Types
Includes all the remaining messages:
* RoadUserChargingReportMessage
* TollUsageMessage
* TollUsageAckMessage
* TrafficLightStatusMessage
* TestMessages 0-15
* Null message types for future use:
  * RoadGeometryAndAttributes
  * PersonalSafetyMessage2
  * SignalControlAndPrioritizationRequest
  * SignalControlAndPrioritizationStatus, 
  * TrafficSignalPhaseAndTiming (TSPAT)

Does not include MessageFrames.

## New: ASN.1 NULL Type Serdes

ASN.1 NULL types are used in two ways in the standard.  The 'future use' messages are defined as aliases of NULL.  For example, PSM2 is defined as:

```asn1
PersonalSafetyMessage2 ::= NULL
```

ASN.1 NULLs are also used in several CHOICE types as a dummy value for flag items, which are fields with no content, but whose presence indicates a choice.   For example the `RptDriverAlertsAndWarnings` type in the `ProbeDataReport` module is defined like:

```asn1
RptDriverAlertsAndWarnings ::= CHOICE {
    j2945Slash1Eebl         NULL,
    j2945Slash1Fcw          NULL,
    j2945Slash1BswLcw       NULL,
    j2945Slash1Ima          NULL,
    j2945Slash1Lta          NULL,
    j2945Slash1Clw          NULL,
    redLightViolationWarn   NULL,
    emerVehAlert            NULL,
    otherAlert              NULL,
    otherWarning            NULL,
    ...
}
```

The chosen item in the type will have a value equal to an ASN.1 NULL.  In JER a NULL is serialized as a JSON `null`, and in XER it is serialized as an empty XML element.  For example, if "redLightViolationWarn" is chosen, the JER serialization of the CHOICE would be:
```json
{
    "redLightViolationWarn": null
}
```
and the XER would be:
```xml
<RptDriverAlertsAndWarnings>
    <redLightViolationWarn/>
</RptDriverAlertsAndWarnings>
```

It is important to handle the serialization of ASN.1 NULLs correctly to be able to represent these CHOICE types.  In the case of JER, it is necessary to be able to distinguish between a property with a null value, and a missing property.  Jackson requires custom serializers/deserializers to accomplish that.

This PR includes new serializers/deserializers to correctly handle ASN.1 NULLs:

* [NullSerializer](https://github.com/CDOT-CV/jpo-asn-pojos/blob/j2735-2024-part-7/jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/serialization/NullSerializer.java)
* [NullDeserializer](https://github.com/CDOT-CV/jpo-asn-pojos/blob/j2735-2024-part-7/jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/serialization/NullDeserializer.java)

Includes unit tests for the "NULL-style" CHOICE types and NULL message types, for example:
* [PersonalSafetyMessage2Test](https://github.com/CDOT-CV/jpo-asn-pojos/blob/j2735-2024-part-7/jpo-asn-j2735-2024/src/test/java/us/dot/its/jpo/asn/j2735/r2024/PersonalSafetyMessage2/PersonalSafetyMessage2Test.java)
* [RptDriverAlertsAndWarningsTest](https://github.com/CDOT-CV/jpo-asn-pojos/blob/j2735-2024-part-7/jpo-asn-j2735-2024/src/test/java/us/dot/its/jpo/asn/j2735/r2024/ProbeDataReport/RptDriverAlertsAndWarningsTest.java)

iyourshaw committed:
RUCRM, TUM, TUMack

iyourshaw committed:
TLSM

iyourshaw committed:
Test and null messages

iyourshaw committed:
Rest of Common

iyourshaw committed:
Merge branch 'j2735-2024-part-6' into j2735-2024-part-7

iyourshaw committed:
Remove print statements from test

iyourshaw committed:
Merge from part 6

iyourshaw committed:
Merge branch 'develop' into j2735-2024-part-7

iyourshaw committed:
Updates from part 9

iyourshaw committed:
Fix TLSM test

iyourshaw committed:
Remove unused

iyourshaw committed:
Null type serdes

iyourshaw committed:
NullDeserializer is working, psm2 test

iyourshaw committed:
Regenerate Asn1Null types

iyourshaw committed:
Tests for null choice types

iyourshaw committed:
Remove unused

mcook42 commented on jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/serialization/NullDeserializer.java:
```diff
@@ -0,0 +1,48 @@
+      throws IOException, JacksonException {
+    // Used by XER only, JER never calls this.
+    // Read the empty node
+    TreeNode node = jsonParser.getCodec().readTree(jsonParser);
+    log.trace("node: {}", node);
```
question(blocking): I don't us using the `node` object anywhere other than in the `log.trace` call. Can we wrap this in a conditional so that we only construct the TreeNode when `trace` log levels are enabled?
```suggestion
    if (log.isTraceEnabled()) {
      TreeNode node = jsonParser.getCodec().readTree(jsonParser);
      log.trace("node: {}", node);
    }
```

iyourshaw replied:
The reason for reading the node, in addition to testing, was that I was thinking the json might need to be "consumed".  But the tests all still pass with this suggestion applied, so apparently it is fine without doing that.

iyourshaw committed:
Update jpo-asn-runtime/src/main/java/us/dot/its/jpo/asn/runtime/serialization/NullDeserializer.java

Co-authored-by: Matt Cook <mattheworion.cook@gmail.com>

payneBrandon commented on jpo-asn-j2735-2024/src/main/java/us/dot/its/jpo/asn/j2735/r2024/TestMessage00/TestMessage00.java:
```diff
@@ -0,0 +1,58 @@
+
+  @Asn1Property(tag = 1, name = "regional", optional = true)
+  @JsonProperty("regional")
+  private Reg_TestMessage00 regional;
```
suggestion: add suppress warnings for rawtypes where needed. Note this applies to all "testmessage##" here

iyourshaw replied:
Added those to all properties of abstract generic types:
[commit: add rawtypes annotations](https://github.com/CDOT-CV/jpo-asn-pojos/pull/19/commits/2577105d820b3c9340092a1f1eee0e5e19cccfe2)

payneBrandon reviewed (changes requested):
Just the one comment so far, looking good

payneBrandon commented:
A couple other questions/suggestions here. I was running into issues with running tests and seeing any results. the 2735 readme has to run `./mvnw clean test` and then the coverage reports are available under 'target/target/site/jacoco/index/html' but this isn't quite correct. The coverage reports _aren't_ generated unless you run the `package` command and then they are available under the `target/site/jacoco/index.html` location. 

There's also an aggregate attempt in here to allow us to view all module tests in a single location, but it doesn't seem to be set up correctly. We need to modify the 2735 pom.xml aggregate section for "includeCurrentProject" to include the boolean value, as it sets to false by default: `<includeCurrentProject>true</includeCurrentProject>`. Then we can view all tests under the one directory. It'd be good to update this to properly aggregate, then update documentation to reflect where to get the aggregate data. 

Not required here, but an interesting note for coverage applications within development environment rather than an html file. If you're using a coverage tool you may need to include these directories in the coverage tool set up since we have this multi-project set configuration. For instance if you use koverage in vscode, you'll need to make the above modification and then additionally configure the settings file for the coverage file path:
```
"koverage.coverageFilePaths": [
        "**/target/site/jacoco-aggregate"
    ]
```

iyourshaw committed:
rawtypes annotations

iyourshaw committed:
Fix Maven test instructions and coverage aggregation

iyourshaw committed:
Node on vscode setting

iyourshaw commented:
> A couple other questions/suggestions here. I was running into issues with running tests and seeing any results. the 2735 readme has to run `./mvnw clean test` and then the coverage reports are available under 'target/target/site/jacoco/index/html' but this isn't quite correct. The coverage reports _aren't_ generated unless you run the `package` command and then they are available under the `target/site/jacoco/index.html` location.
> 
> There's also an aggregate attempt in here to allow us to view all module tests in a single location, but it doesn't seem to be set up correctly. We need to modify the 2735 pom.xml aggregate section for "includeCurrentProject" to include the boolean value, as it sets to false by default: `<includeCurrentProject>true</includeCurrentProject>`. Then we can view all tests under the one directory. It'd be good to update this to properly aggregate, then update documentation to reflect where to get the aggregate data.
> 
> Not required here, but an interesting note for coverage applications within development environment rather than an html file. If you're using a coverage tool you may need to include these directories in the coverage tool set up since we have this multi-project set configuration. For instance if you use koverage in vscode, you'll need to make the above modification and then additionally configure the settings file for the coverage file path:
> 
> ```
> "koverage.coverageFilePaths": [
>         "**/target/site/jacoco-aggregate"
>     ]
> ```

Addressed these issues (Fixed the maven instructions in the readme, fixed agg test coverage config, and added a note on that vscode setting)

payneBrandon reviewed (approved):
lgtm!
